app/models/user.rb

```diff
@@ -247,30 +344,39 @@ def self.new_with_session(params, session)
 
   # Teleports to another user
   def teleport(user)
-    ActionCable.server.broadcast("location_#{location_id}", method: 'player_warp_out', name: full_name)
+    self.location.broadcast(:player_warp_out, name: full_name)
     old_system = self.system
-    self.update_columns(location_id: user.location_id, system_id: user.system_id, docked: user.docked, in_warp: false)
+    self.update(location_id: user.location_id, system_id: user.system_id, docked: user.docked, in_warp: false)
     # Tell everyone in old system to update their local players
     old_system.update_local_players
     # Tell everyone in new system to update their local players
     self.reload.system.update_local_players
-    ActionCable.server.broadcast("location_#{location_id}", method: 'player_appeared')
-    ActionCable.server.broadcast("player_#{id}", method: 'warp_finish')
+    self.location.broadcast(:player_appeared)
+    self.broadcast(:warp_finish)
   end
 
-  # Ban User
   def ban(duration, reason)
     if duration.to_i == 0
-      self.update_columns(banned: true, banned_until: nil, banreason: reason)
+      self.update(banned: true, banned_until: nil, banreason: reason)
     else
-      self.update_columns(banned: true, banned_until: (DateTime.now.to_time + duration.to_i.hours).to_datetime , banreason: reason)
+      self.update(banned: true, banned_until: (DateTime.now.to_time + duration.to_i.hours).to_datetime , banreason: reason)
     end
-    ActionCable.server.broadcast("player_#{id}", method: 'reload_page')
+    self.broadcast(:reload_page)
   end
 
-  # Unban User
   def unban
-    self.update_columns(banned: false, banned_until: nil, banreason: nil) if banned
+    self.update(banned: false, banned_until: nil, banreason: nil)
+  end
+
+  def has_blueprints_for?(loader)
+    self.blueprints.where(loader: loader).exists?
+  end
+
+  def give_blueprints_for(loader, efficiency: 1.5)
+    blueprint = Blueprint.where(user: self, loader: loader).first_or_initialize
+    blueprint.efficiency = efficiency
+    blueprint.save
+    blueprint
```

Comment:
doesn't `blueprint.save` already return the blueprint?

Reply:
true... but only if the record is `valid` -- otherwise it returns `false`

Reply:
we should maybe rescue if the record is invalid

Reply:
calling `.save!` instead -- which throws an exception if the record is invalid